mitiq/shadows/shadows.py

```diff
@@ -0,0 +1,130 @@
+# Copyright (C) Unitary Fund
+#
+# This source code is licensed under the GPL license (v3) found in the
+# LICENSE file in the root directory of this source tree.
+"""Classical shadow estimation for quantum circuits. Based on the paper"""
+
+from typing import Optional, Callable, List, Dict, Any
+
+import cirq
+import numpy as np
+from numpy.typing import NDArray
+
+import mitiq
+from mitiq import MeasurementResult
+from mitiq.shadows import (
+    random_pauli_measurement,
+    shadow_state_reconstruction,
+    expectation_estimation_shadow,
+)
+from mitiq.shadows.shadows_utils import (
+    n_measurements_tomography_bound,
+    n_measurements_opts_expectation_bound,
+)
+
+
+def execute_with_shadows(
+    circuit: cirq.Circuit,
+    executor: Callable[[cirq.Circuit], MeasurementResult],
+    observables: Optional[List[mitiq.PauliString]] = None,
+    state_reconstruction: bool = False,
+    *,
+    k_shadows: Optional[int] = None,
+    num_total_measurements: Optional[int] = None,
+    error_rate: Optional[float] = None,
+    failure_rate: Optional[float] = None,
+    random_seed: Optional[int] = None,
+) -> Dict[str, NDArray[Any]]:
+    r"""
+    Executes a circuit with classical shadows. This function can be used for
+    state reconstruction or expectation value estimation of observables.
+
+    Args:
+        circuit: The circuit to execute.
+        executor: The function to use to do quantum measurement.
+        observables: The set of observables to measure. If None, the state
+            will be reconstructed.
+        state_reconstruction: Whether to reconstruct the state or estimate
+            the expectation value of the observables.
+        k_shadows: Number of groups of "median of means" used for shadow
+            estimation.
+        num_total_measurements: Number of shots per group of
+            "median of means" used for shadow estimation.
+        num_total_measurements: Total number of shots for shadow estimation.
+        error_rate: Predicting all features with error rate
+            :math:`\epsilon` via median of means prediction.
+        failure_rate: :math:`\delta` Accurately predicting all features via
+            median of means prediction with error rate less than or equals to
+            :math:`\epsilon` with probability at least :math:`1 - \delta`.
+        random_seed: The random seed to use for the shadow measurements.
+
+    Returns:
+        A dictionary containing the shadow outcomes, the Pauli strings, and
+        either the estimated density matrix or the estimated expectation
+        values of the observables.
+    """
+
+    qubits = list(circuit.all_qubits())
+    num_qubits = len(qubits)
+
+    if observables is None:
+        assert (
```

Comment:
Instead of using `assert`, it is better to use `ValueError`, `TypeError`, or `RuntimeError`, etc... 

If I am not wrong, assert statement should be used for statements that are _always_ `True`, unless there is a bug in the code. This is why they are typically used in tests.

For errors that are _possible_ even with a correct code without bugs (due to mistakes of the user) rising Python errors is the best practice.